### DISCUSSION TITLE
How can I set a key/resolver for nested data with filters?

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#2205] [Q&A] [answered]
I'm trying to figure out how to follow the pagination example from the docs. I think the part that the docs are missing is how to deal with id/key fields for nested items that also have filters.

Here is my query:

```graphql
query VenueCalendarBookings(
  $slug: String!
  $start: DateTime!
  $offset: Int!
  $limit: Int!
) {
  venue(slug: $slug) {
    id
    location {
      timezone
    }
    calendarBookings(start: $start, limit: $limit, offset: $offset) {
      items {
        id
      }
      canFetchMore
    }
  }
}
```

The `calendarBookings` returns a `CalendarBookingsResponse` type. My first instinct was to set the `keys` to null for this type:

```ts
cacheExchange({ keys: { CalendarBookingsResponse: () => null } })
```

In principle, this seemed fine, since `calendarBookings` is nested inside of a venue and can be an embedded entity.

However, this isn't desired. The result is, every time I expand to another page, all existing pages refetch. This makes sense, as they have no unique identifiers from the `args`.

So what I really want is for `calendarBookings` to be identified by the combination of `args` and `venueId`. My next best guess is the resolvers. However, I can't find a good example for the resolvers from the docs. I see examples for resolving nested fields, and also for resolving entities based on their id, but **I'm not sure how to resolve a list of entities based on an ID + args.**

Any help here would be great, thank you.

**Comment by JoviDeCroock** [comment, marked as answer]
I think we could be looking at it a bit wrong here, it's similar to another discussion earlier, we're dealing with an edge here but you're mixing node and edge data, i.e. your data is showing signs of `fetchMore` mixed with data-attributes.

Secondly, an edge we wouldn't want to uniquely identify but a node we would, hence a  structure like

```
    calendarBookings(start: $start, limit: $limit, offset: $offset) {
      __typename // CalendarEdge
      nodes { venueLocation ... }
      pageInfo { canFetchMore }
    }
```

This can also be done under the numeration of a simple-pagination but as long as we don't mix data-concerns we can cleanly key off.
That being said in the background the following will happen `Venue:id.calendarBookings(yourArgs) --> TheResponse`

**Reply by nandorojo** [reply]
This makes sense to me, although from what I can tell it appears that these are mostly naming preferences. Our current `calendarBookings` type is essentially acting as an edge type, right?

> That being said in the background the following will happen `Venue:id.calendarBookings(yourArgs) --> TheResponse`

This is actually desired, which is why I'm confused by the re-fetching when I create a new page. I'm glad you clarified that, because this is what I figured the behavior would be. However, every time I create a new page, it refetches all existing pages.

Does this behavior maintain itself if I set the key for `CalendarEdge` to `null`?

**Reply by nandorojo** [reply]
I re-organized some fields around in the original query to clarify the layout.

**Reply by nandorojo** [reply]
(And FWIW, my codegen is injecting `__typename` into each entity)

**Reply by JoviDeCroock** [reply]
Made a small example 😅 https://codesandbox.io/s/sleepy-morning-nl5dc?file=/src/components/Todos.js it does seem to cache appropriately for me atleast. Clicking prev - next just works

**Reply by nandorojo** [reply]
You're right. In fact, I made the same example here with our real data, and it worked: https://codesandbox.io/s/lively-framework-pqy3z?file=/src/index.js

What the example is lacking, however, is the infinite scroll component, where all pages are rendered at the same time. I'm going to see if the behavior is preserved when that's the case.

**Reply by nandorojo** [reply]
Here I added infinite scrolling, and it doesn't appear to have any duplicate request issues.

https://codesandbox.io/s/infinite-01mkr?file=/src/components/Todos.js

I think the problem was the `ttl` I set:

```ts
requestPolicyExchange({
  ttl: 3000,
})
```

Once I removed the `ttl`, it seems to have worked. Sorry about that, still getting a hang of all the different config options.
